hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/ConsensusSubmitMessageHandler.java

```diff
@@ -68,22 +89,162 @@ public void preHandle(@NonNull final PreHandleContext context, @NonNull Readable
+     * Throws {@link HandleStatusException} if any of the validations fail.
+     * @param txnId the {@link TransactionID} of the active transaction
+     * @param payer the {@link AccountID} of the payer
+     * @param op the {@link ConsensusSubmitMessageTransactionBody} of the active transaction
+     */
+    private void validateChunkInfo(
+            final TransactionID txnId, final AccountID payer, final ConsensusSubmitMessageTransactionBody op) {
+        if (op.hasChunkInfo()) {
+            var chunkInfo = op.getChunkInfo();
+
+            /* Validate chunk number */
+            if (!(1 <= chunkInfo.getNumber() && chunkInfo.getNumber() <= chunkInfo.getTotal())) {
+                throw new HandleStatusException(INVALID_CHUNK_NUMBER);
+            }
+
+            /* Validate the initial chunk transaction payer is the same payer for the current transaction*/
+            if (!chunkInfo.getInitialTransactionID().getAccountID().equals(payer)) {
+                throw new HandleStatusException(INVALID_CHUNK_TRANSACTION_ID);
+            }
+
+            /* Validate if the transaction is submitting initial chunk,payer in initial transaction Id should be same as payer of the transaction */
+            if (1 == chunkInfo.getNumber()
+                    && !chunkInfo.getInitialTransactionID().equals(txnId)) {
+                throw new HandleStatusException(INVALID_CHUNK_TRANSACTION_ID);
+            }
+        }
+    }
+
+    /**
+     * Updates the running hash and sequence number of the topic.
+     * @param txn the {@link TransactionBody} of the active transaction
+     * @param topic the topic to which the message is being submitted
+     * @param consensusNow the consensus time of the active transaction
+     * @return the updated topic
+     * @throws IOException if there is an error while updating the running hash
+     */
+    public Topic updateRunningHashAndSequenceNumber(final TransactionBody txn, final Topic topic, Instant consensusNow)
+            throws IOException {
+        final var payer = txn.getTransactionID().getAccountID();
+        final var topicId = txn.getConsensusSubmitMessage().getTopicID();
+        final var message = txn.getConsensusSubmitMessage().getMessage().toByteArray();
+
+        // This line will be uncommented once there is PBJ fix to make copyBuilder() public
+        // final var topicBuilder = topic.copyBuilder();
+        final var topicBuilder = new Topic.Builder()
+                .topicNumber(topic.topicNumber())
+                .adminKey(topic.adminKey())
+                .submitKey(topic.submitKey())
+                .autoRenewAccountNumber(topic.autoRenewAccountNumber())
+                .autoRenewPeriod(topic.autoRenewPeriod())
+                .expiry(topic.expiry())
+                .deleted(topic.deleted())
+                .memo(topic.memo());
+
+        if (null == consensusNow) {
+            consensusNow = Instant.ofEpochSecond(0);
```

Comment:
Why are we assigning a timestamp of jan 1 1970 here?

Reply:
I think there are cases when platform consensus timestamp can be null. I am not very sure. @tinker-michaelj Can we remove this check ?